compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -104,12 +104,11 @@ fn map_error<'tcx>(
             // This is sometimes not a compile error if there are trivially false where clauses.
             // See `tests/ui/layout/trivial-bounds-sized.rs` for an example.
             assert!(field.layout.is_unsized(), "invalid layout error {err:#?}");
-            if !field.ty.is_sized(cx.tcx(), cx.typing_env) {
```

Comment:
I would really prefer if we found some way to assert that there actually are unsatisfiable predicates in the environment here... At minimum this should assert that there are some predicates at all, i.e. `if cx.typing_env.param_env.caller_bounds.is_empty() { delayed_bug }`.

Also, this assertion is the only reason why the `field` of `LayoutCalculatorError::UnexpectedUnsized` exists and if we end up removing the assert, then we can also get rid of the field as well as the `without_payload` machinery here: https://github.com/rust-lang/rust/blob/243d2ca4db6f96d2d18aaf3a2381251d38eb6b0b/compiler/rustc_abi/src/layout.rs#L65 (this can be a follow-up and doesn't have to happen in this PR).

Reply:
I was not aware of this, due to my prematurity.
Thanks, I'll look for something like this.